### ISSUE TITLE
Simple OAuth2 Facebook authentication freezing UIWebView

### ISSUE THREAD

**Issue by ammarMheir** [xamarin/Xamarin.Auth#144]
##Issue Defniniton
Creating a new single view Xamarin.iOS application and adding the following code will hang the UIWebView right after inputting username and password (on page asking for permissions):

##Steps to replicate
1) go to https://developers.facebook.com/
2) Create a New app
3) Copy Client ID
4) Use following code with newly create Client ID

var auth = new OAuth2Authenticator(
                clientId: "REPLACE WITH CLIENT ID",
                scope: "",
                authorizeUrl: new Uri("https://m.facebook.com/dialog/oauth/"),
                redirectUrl: new Uri("http://www.facebook.com/connect/login_success.html"));

            auth.AllowCancel = true;
            auth.Completed += Auth_Completed;
            PresentViewController(auth.GetUI(), true, null);

Was just following exactly how it was being setup using the "Getting Started" page. The authenticator will run as expected however as soon as pressing the login button and the next page appears as so: http://imgur.com/a/ptxg3 

The UIWebView will become unresponsive not allowing any sort of interaction. However canceling still works.

The same login is being used on Android with no issues at all. Tested on physical device as well.

##Version Info
Microsoft Visual Studio Enterprise 2015
Version 14.0.25431.01 Update 3
Microsoft .NET Framework
Version 4.6.01586

Xamarin   4.2.2.11 (00fa5cc)
Xamarin.iOS   10.3.1.8 (7beaef4)

**Comment by ammarMheir** [comment]
Just found issue [#108](https://github.com/xamarin/Xamarin.Auth/issues/108) and was wondering if this also applies for this error? Although It is working for Android. Do I need to use the Facebook SDK for Xamarin.iOS apps?

**Comment by sushihangover** [comment]
@ammarMheir  Item no. 2 of https://developers.facebook.com/policy/#login

>Android apps should use the default login behavior defined by the SDK, which may use the web-view Login dialog. On iOS, only kiosk apps may use a web-view Login dialog.

You will be told by Facebook that your app is not in compliance with Facebook's login requirements when going through their login approval process if you are using `Xamarin.Auth` on iOS... for right now Android-based `WebView` is still allowed, but there are rumors that Chrome CustomTabs will be a future requirement (just a rumor right now, they could require using their native Android SDK instead...) I went native SDK on both platforms for multiple apps now and Facebook has never "rejected" any permissions that we have requested.

Personally, I dropped usage of `Xamarin.Auth` as it is does not support iOS' SFSafariViewController and Android Chrome CustomTabs. Starting next month (3/17) Google will start rejecting apps that do not support using Chrome CustomTabs for Google logins on devices that support an alternative to `WebView` (basically any device that ships with a version of Chrome pre-installed, which now including almost all devices in the USA, Canada, most countries in Europe, etc... China is one of the few areas that Chrome is not included in the ROM as the user preferences lean toward in-country browser's such as Baidu Explorer and using a Google login is not high on their login requirements.... ;-)

**Comment by leongaines** [comment]
@ammarMheir 

i am experiencing the same issue.  What I noticed is that after entering user name and password it will say "Continue as [Name]" but the button is not clickable.  I noticed that only after I log into the mobile app does the button become enabled.  Strange behavior.  I cannot guarantee that my users are using the mobile app.  This is occuring on iOS6.  However, in debug mode using the iPhone Simulator I am able to authenticate properly.    Does this sound similar to what you're experiencing?

**Comment by hvaughan3** [comment]
@leongaines Same thing I have seen as well. Everyone now and then it would work and then break again. One time I changed the return URL and it started working but eventually broke again. Like sushihangover said, Facebook no longer wants apps using a WebView for logging in. You could try sushihangover's suggestion but I might try using the native Facebook app (if it is available) by following the link below. Just need to find some time to try it out!

[https://forums.xamarin.com/discussion/31928/xamarin-forms-facebook-login-native-login-page-android](https://forums.xamarin.com/discussion/31928/xamarin-forms-facebook-login-native-login-page-android)

**Comment by ammarMheir** [comment]
@leongaines 

Indeed this is the exact same thing that is happening with me. What @sushihangover is correct and it seems that not just Facebook are doing this but also Google+ (pretty soon). Embedded WebViews apparently have some security issues and will not work well in authenticating with these social media providers.

However what I found is that if you download the "[portable-bait-and-switch](https://github.com/xamarin/Xamarin.Auth/tree/portable-bait-and-switch)" branch from GitHub, it has an updated version of the authenticator on iOS that uses the SFSafariViewController rather than a UIWebView. Which is essentially opening the web browser to do the authentication on top of the app and it seems to be working fine using the browser rather than the WebView. Although this works, i still think it might be best to go with the Facebook SDK as it might provide more stability in the long run. 

Thanks @sushihangover and @hvaughan3 for the info on this.

**Comment by hvaughan3** [comment]
@ammarMheir One more thing I would say, the link I posted in the previous message will actually fallback to using `SFSafariViewController` if the user does not have the Facebook app installed on their device. Facebook also suggests doing this type of fallback so, at least in the near future, I think using the `SFSafariViewController` is a good approach... until they change everything on us again -.-

**Comment by leongaines** [comment]
@ammarMheir  @hvaughan3 ,

Here is what i've come up with.  If you use the native app for authentication, but the user has never logged in with a native app you will still have the same issue.  Try it.  for the account that you're having trouble with, log into the native app with it, then try your web view.  It works in my case every time.   The problem isn't necessarily the webview.  It's a business decision by Facebook that if authentication is via a mobile app then the user must have used their Facebook mobile app.  Keep in mind, iOS, your Facebook account is tightly tethered to your iPhone (Settings: Facebook: Settings : account).  

Bottom line, mobile apps that require FB authentication needs to have logged into the mobile app to be provisioned.

**Comment by moljac** [comment]
Hi

[Chrome] Custom Tabs and SFSafariViewController are under heavy development. 

For more details please check https://github.com/xamarin/Xamarin.Auth/issues/150

**Comment by hvaughan3** [comment]
@moljac Very exciting, thank you!

**Comment by sushihangover** [comment]
@moljac Just an FYI, Google has already started blocking new apps that are only providing a `WebView` OAuth2 option as the primary browser option. This is ahead of their originally 3/17/2017 ban/deadline, but after talking to support about one of my client's apps, they pointed out that October 20, 2016 was the initial offcut for new clients not supporting Chrome CustomTabs and the "updated" April 20, 2017 date is the official ban date for all clients using WebView only on Android....

**Comment by moljac** [comment]
@sushihangover 
The official announcement was given 2016-08-22

https://developers.googleblog.com/2016/08/modernizing-oauth-interactions-in-native-apps.html

Personally I experienced "problems" with new Apps on 2016-10-28 and preventing for new apps started on 2016-10-20.  After that we received several bugreports and issues regarding new apps and Xamarin.Auth.  I recall that I was a bit surprised with ban in October, but could not recall 2017-03-17. This is the reason I got caught off guard with this.

**Comment by leongaines** [comment]
The answer is to use Xamarin.Social framework!

**Comment by sinclas** [comment]
Hi,

I have an app experience this problem.  Should I use Xamarin.Soclal from now on?   What is the official work-around.  Stop using this component?

thanks!

**Comment by hvaughan3** [comment]
@sinclas Update above version 1.5.0 and set `isUsingNativeUI ` to `true`

**Comment by sinclas** [comment]
How do I eliminate the schema warning?   I remeber finding some references to that issue in Git but can’t find it.  I have the isUsingNativeUI set to true.   

-Thanks!
Steve

**Comment by moljac** [comment]
@sinclas 

Nullify it:

https://github.com/xamarin/Xamarin.Auth/blob/master/samples/Traditional.Standard/Providers/Xamarin.Auth.Sample.XamarinAndroid/MainActivity.cs#L125-L127

**Comment by kggp1995** [comment]
So guys, we are stuck at the "Authenticate" web view (i.e., FB consent screen) once the user successfully logs in. I am now aware that FB is inhibiting auth via WebView. But what workaround does Xamarin offer?

All help and references are appreciated!
TIA
  _
<X>
  ‾‾

**Comment by newky2k** [comment]
@ammarMheir @sushihangover @sinclas @hvaughan3 @leongaines can you check to see if this is still happening in the latest version.

Regards

**Comment by newky2k** [comment]
no response after two weeks closing
